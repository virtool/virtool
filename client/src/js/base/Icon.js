import React, { useCallback } from "react";
import PropTypes from "prop-types";
import { capitalize, get } from "lodash-es";
import { Link } from "react-router-dom";
import styled, { css } from "styled-components";
import { Tooltip } from "./Tooltip";

const getIconColor = ({ color, theme, shade = "dark" }) =>
    get(theme, ["color", `${color}${capitalize(shade)}`], "inherit");

const fixedWidth = css`
    width: 8px;
    text-align: center;
    display: inline-block;
`;

const StyledIcon = styled.i`
    color: ${getIconColor};
    ${props => (props.hoverable || props.onClick ? "cursor: pointer;" : "")};
    opacity: ${props => (props.hoverable || props.onClick ? 0.7 : 1)};

    ${props => (props.fixedWidth ? fixedWidth : "")};

    :hover {
        opacity: 1;
    }
`;

export const Icon = ({ hoverable, style, ...props }) => {
    const handleClick = useCallback(e => {
        props.onClick(e);
    }, []);

    const className = `${props.className ? props.className + " " : ""} ${props.faStyle} fa-${props.name}`;

    const icon = (
        <StyledIcon
            className={className}
            fixedWidth={props.fixedWidth}
            hoverable={hoverable}
            style={style}
            onClick={props.onClick ? handleClick : null}
            color={props.color}
            shade={props.shade}
        />
    );

    if (props.tip) {
        return (
            <Tooltip position={props.tipPlacement || "top"} tip={props.tip}>
                {icon}
            </Tooltip>
        );
    }

    return icon;
};

Icon.propTypes = {
    color: PropTypes.oneOf(["blue", "green", "grey", "red", "orange", "purple"]),
    name: PropTypes.string.isRequired,
    tip: PropTypes.node,
    tipPlacement: PropTypes.oneOf(["top", "right", "bottom", "left"]),
    faStyle: PropTypes.string,
    onClick: PropTypes.func,
    className: PropTypes.string,
    fixedWidth: PropTypes.bool,
    style: PropTypes.object
};

Icon.defaultProps = {
    faStyle: "fas",
    fixedWidth: false
};

export const LinkIcon = ({ to, replace, ...props }) => (
    <Link to={to} replace={replace} hoverable>
        <Icon {...props} />
    </Link>
);
