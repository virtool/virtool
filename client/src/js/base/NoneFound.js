import React from "react";
import PropTypes from "prop-types";
import { ListGroup, ListGroupItem } from "react-bootstrap";
import { Icon } from "./Icon";

/**
 * A ListGroupItem component with a 'none found'-type message. Used in ListGroups when no data is available to populate
 * the list. For example, when no sample have been created.
 *
 * @param noun {string} the name of the items of which none were found (eg. samples)
 * @param noListGroup {boolean} don't include a ListGroup in the returned element
 */
export const NoneFound = ({ noun, noListGroup }) => {
    const item = (
        <ListGroupItem className="text-center">
            <Icon name="info" /> No {noun} found
        </ListGroupItem>
    );

    if (noListGroup) {
        return item;
    }

    return (
        <ListGroup>
            {item}
        </ListGroup>
    );
};

NoneFound.propTypes = {
    noun: PropTypes.string.isRequired,
    noListGroup: PropTypes.bool
};
