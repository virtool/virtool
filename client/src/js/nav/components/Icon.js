import React from "react";
import { connect } from "react-redux";
import { Overlay } from "react-bootstrap";
import Notifications from "./Notifications";
import { getSoftwareUpdates } from "../../updates/actions";
import { getUnbuilt } from "../../indexes/actions";

import { Icon } from "../../base";

const hrToMs = 3600000;

class NotificationIcon extends React.Component {

    constructor (props) {
        super(props);

        this.state = {
            show: false
        };
    }

    handleToggle = () => {

        if (this.state.show) {
            window.removeEventListener("click", this.handleExit, false);
        } else {
            window.addEventListener("click", this.handleExit, false);
        }

        this.setState({
            show: !this.state.show
        });
    }

    handleExit = (e) => {
        if (!this.target.contains(e.target)) {
            this.handleToggle();
        }
    }

    componentWillMount () {
        this.props.onGet();
        this.props.onGetUnbuilt();
    }

    componentDidMount () {
        this.interval = window.setInterval(() => {
            this.props.onGet();
            this.props.onGetUnbuilt();
        }, (3 * hrToMs));
    }

    componentWillUnmount () {
        window.clearInterval(this.interval);
    }

    render () {

        const availableUpdates = this.props.updates ? this.props.updates.releases.length : null;
        const unbuiltIndex = ((this.props.modifiedVirusCount || this.props.modifiedCount) && this.props.unbuilt)
            ? this.props.unbuilt.history.length
            : null;

        const iconStyle = (availableUpdates || unbuiltIndex) ? "icon-pulse" : "icon";

        return (
            <div>
                <div ref={node => this.target = node} onClick={this.handleToggle}>
                    <Icon
                        className={iconStyle}
                        name="notification"
                        tip="Click to see notifications"
                        tipPlacement="left"
                    />
                </div>

                <Overlay
                    show={this.state.show}
                    placement="bottom"
                    target={this.target}
                >
                    <Notifications onClick={this.handleToggle} />
                </Overlay>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    updates: state.updates.software,
    unbuilt: state.indexes.unbuilt,
    modifiedVirusCount: state.indexes.modified_virus_count,
    modifiedCount: state.viruses.modified_count
});

const mapDispatchToProps = (dispatch) => ({

    onGet: () => {
        dispatch(getSoftwareUpdates());
    },

    onGetUnbuilt: () => {
        dispatch(getUnbuilt());
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(NotificationIcon);
