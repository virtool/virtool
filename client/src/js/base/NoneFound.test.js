import { NoneFound } from "./NoneFound";
import { ListGroup, ListGroupItem } from "react-bootstrap";
import { Icon } from "./Icon";

describe("<NoneFound />", () => {
    let wrapper;

    it("renders an info Icon component", () => {
        const noun = "test";

        wrapper = shallow(<NoneFound noun={noun} />);

        expect(wrapper.find(Icon).exists()).toBe(true);
        expect(wrapper.find(Icon).prop('name')).toEqual("info");
        expect(wrapper).toMatchSnapshot();
    });

    describe("when supplied [noListGroup=false] prop", () => {
        const noun = "test";
        const noListGroup = false;

        beforeEach(() => {
            wrapper = shallow(<NoneFound noun={noun} noListGroup={noListGroup} />);
        });

        it("renders correctly", () => {
            expect(wrapper).toMatchSnapshot();
        });

        it("renders a ListGroup", () => {
            expect(wrapper.find(ListGroup).exists()).toBe(true);

            const message = wrapper.find(ListGroupItem).childAt(1).text() +
                wrapper.find(ListGroupItem).childAt(2).text() +
                wrapper.find(ListGroupItem).childAt(3).text();
            expect(message).toEqual(` No ${noun} found`);
        });

    });

    describe("when supplied [noListGroup=true] prop", () => {
        const noun = "test";
        const noListGroup = true;

        beforeEach(() => {
            wrapper = shallow(<NoneFound noun={noun} noListGroup={noListGroup} />);
        });

        it("renders correctly", () => {
            expect(wrapper).toMatchSnapshot();
        });

        it("renders a ListGroupItem without ListGroup container", () => {
            expect(wrapper.find(ListGroupItem).exists()).toBe(true);
        });
    });

});
