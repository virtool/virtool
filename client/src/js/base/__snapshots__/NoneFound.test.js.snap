// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoneFound /> renders an info Icon component 1`] = `
<ListGroup
  bsClass="list-group"
>
  <ListGroupItem
    bsClass="list-group-item"
    className="text-center"
    listItem={false}
  >
    <Icon
      fixedWidth={false}
      name="info"
      pending={false}
      pullRight={false}
    />
     No 
    test
     found
  </ListGroupItem>
</ListGroup>
`;

exports[`<NoneFound /> when supplied [noListGroup=false] prop renders correctly 1`] = `
<ListGroup
  bsClass="list-group"
>
  <ListGroupItem
    bsClass="list-group-item"
    className="text-center"
    listItem={false}
  >
    <Icon
      fixedWidth={false}
      name="info"
      pending={false}
      pullRight={false}
    />
     No 
    test
     found
  </ListGroupItem>
</ListGroup>
`;

exports[`<NoneFound /> when supplied [noListGroup=true] prop renders correctly 1`] = `
<ListGroupItem
  bsClass="list-group-item"
  className="text-center"
  listItem={false}
>
  <Icon
    fixedWidth={false}
    name="info"
    pending={false}
    pullRight={false}
  />
   No 
  test
   found
</ListGroupItem>
`;
